airflow/api_connexion/endpoints/task_instance_endpoint.py

```diff
@@ -61,13 +61,8 @@
 T = TypeVar("T")
 
 
-@security.requires_access(
-    [
-        (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG),
-        (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG_RUN),
-        (permissions.ACTION_CAN_READ, permissions.RESOURCE_TASK_INSTANCE),
-    ],
-)
+@security.requires_access_dag("GET", DagAccessEntity.RUN)
+@security.requires_access_dag("GET", DagAccessEntity.TASK_INSTANCE)
```

Comment:
Finally got to review it after the break (and this and next week will have a lot of focus on those).

I think this really calls for small refactoring: 

```python
def requires_access_dag(
    method: ResourceMethod, access_entities: tuple[DAGAccessEntity] | DagAccessEntity | None = None
```

and then:

```
@security.requires_access_dag("GET", (DagAccessEntity.RUN, DagAccessEntity.TASK_INSTANCE))
```

and convert `access_entity` to `access_entities: tuple[DAGDetails]` (we can convert single dag detail to one-element tuple).

Not sure if now or as a follow up to this PR, but I am quite sure this will be much better for performance point of view (one callback instead of two and it will be pretty much impossible to optimize it if we have two decorators. 

And since it will change the callback, it's likely better to do it it now.

Reply:
Looking at the comments I looked up below I have some doubts and different idea.

I think the original permission for task endpoints were just wrong:

```python
@security.requires_access(
    [
        (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG),
        (permissions.ACTION_CAN_READ, permissions.RESOURCE_TASK_INSTANCE),
    ],
)
```

It really should be:

```python
@security.requires_access(
    [
        (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG),
        (permissions.ACTION_CAN_READ, permissions.RESOURCE_TASK),
    ],
)
```

I know we do not have "RESOURCE_TASK" - but this should be really how it should be done inititially. Those endpoints do not access task instance nor dag_run access - they just add `task` definition. I think we wrongly had only one `task_instance` resource where we should have two different resources for those.

IMHO  - we have the opportunity to fix it (and that will allow us to avoid adding tuple handling from my above proposal.

Why don't we simplify it all by 

* `DagAccessEntity.DAG`
* `DagAccessEntity.TASK`
* `DagAccessEntity.RUN`
* `DagAccessEntity.TASK_INSTANCE` 

We should only need to specify TASK_INSTANCE entity. In FAB implementation we could map:

* DagEntity.TASK => resource.TASK_INSTANCE and 
* DagEntity.TASK_INSTANCE -> resource.DAG_RUN + resource.TASK_INSTANCE

The fact that sometimes we specified "read" for DAG_RUN and "write" to "TASK_INSTANCE" resources was I think a mistake. I can't imagine situation where we would like to limit the user from writing to DAG_RUN where the user should be able to update TASK_INSTANCE. So this would automatically fix the bad permissions we had originally for those endpoints.

Reply:
BTW. This also reflects the change in the new approach of the permission model after AIP-56 implementation. We are no more "resource" driven. We are "action" driven. Auth manager should not be checking if you are access to specrific **resources**, it instead checks if you are allowed to execute specific **action**. 

Rather than enumerating the resources necessary to execute the action, auth manager simply responds to the question "can this user execute this action?". For example:

* Is the user able to "modify" this task_instance (and all that it implies)?
* Is the user able to "read" this dag run (and all that it implies)?

Reply:
Strongly agree on this one. I tried to achieve this and as you said, just declare the action the user is trying to do, then the auth manager is responsible of checking the permissions needed to perform such action. The new resource type `TASK` makes and simplify it. I dont like having to check `DAG_RUN` and `TASK_INSTANCE` at the same time.

> Why don't we simplify it all by
> 
> DagAccessEntity.DAG
> DagAccessEntity.TASK
> DagAccessEntity.RUN
> DagAccessEntity.TASK_INSTANCE

My approach is, if no `DagAccessEntity` is defined then it is on the Dag level. My thinking is `DagAccessEntity` is here to give more information on which entity/information of a DAG the user is trying to access. If none is specified, then it is on the DAG itself. You rather being explicit here?

Also, just to confirm. We still need the other values in `DagAccessEntity` right? Such as `AUDIT_LOG`, `TASK_LOGS`, ...

Reply:
I added `DagAccessEntity.TASK`